File: skypydb-js/src/api/mixins/vector/create_collection.ts
import type { Metadata } from "../../../types";
import { Collection } from "../../collection";
import type { VectorDatabase } from "../../../database/vector_db";

export class CreateCollectionMixin {
  protected _db!: VectorDatabase;
  protected _collections!: Map<string, Collection>;

  async create_collection(
    name: string,
    metadata?: Metadata,
    get_or_create = false
  ): Promise<Collection> {
    if (get_or_create) {
      const current = this as unknown as {
        get_or_create_collection: (name: string, metadata?: Metadata) => Promise<Collection>;
      };
      await current.get_or_create_collection(name, metadata);
    } else {
      this._db.create_collection(name, metadata);
    }

    const existing = this._collections.get(name);
    if (existing) {
      return existing;
    }

    const collection = new Collection(this._db, name, metadata ?? {});
    this._collections.set(name, collection);
    return collection;
  }
}

File: skypydb/api/operations/vector/create_collection.py
"""Mixin that creates vector collections for `VectorClient`."""

from typing import Any, Dict, Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from skypydb.api.collection import Collection


class CreateCollectionMixin:
    """Expose collection creation behavior on `VectorClient`."""

    def create_collection(
        self,
        name: str,
        metadata: Optional[Dict[str, Any]] = None,
        get_or_create: bool = False,
    ) -> "Collection":
        """Create a new collection and return a cached `Collection` wrapper."""

        if get_or_create:
            self.get_or_create_collection(name, metadata)
        else:
            self._db.create_collection(name, metadata)

        collection = self._collections.get(name)
        if collection is None:
            from skypydb.api.collection import Collection

            collection = Collection(db=self._db, name=name, metadata=metadata)
            self._collections[name] = collection
        return collection
